crates/commitlog/src/dio/writer.rs

```diff
@@ -0,0 +1,214 @@
+use std::{
+    cmp,
+    io::{self, Write},
+};
+
+use log::warn;
+
+use crate::segment::FileLike;
+
+use super::{
+    page::{Page, BLOCK_SIZE, PAGE_SIZE},
+    WriteAt,
+};
+
+/// A buffered writer using an aligned buffer internally.
+///
+/// Similar to [`io::BufWriter`], but suitable for files opened using `O_DIRECT`
+/// or a platform equivalent, due to the alignment.
+///
+/// # Flushing behaviour
+///
+/// [`io::Write::write`] calls will only flush the buffer when it is full.
+/// [`io::Write::flush`] calls, however, will flush the buffer up to the next
+/// [`BLOCK_SIZE`] boundary, padding the data with zeroes if necessary.
+///
+/// This is done so that partial writes to the underlying storage can be
+/// detected on the application layer, and retried if appropriate. It is also
+/// necessary to preserve `fsync` semantics: a [`PagedWriter`] replaces the OS
+/// page cache, where the latter is flushed to the device when `fsync` is called.
```

Comment:
Challenge this.

It seemed confusing having to remember to flush first, even though `BufWriter` behaves like this -- except that a flush there only means to move the buffered data to kernel space.

Reply:
I think this makes sense for our use case. What exactly is the alternative?